File: python/src/kanbus/standup_window.py
"""Standup time window resolution (rolling vs calendar)."""

from __future__ import annotations

import os
import re
from dataclasses import dataclass
from datetime import date, datetime, time, timedelta, timezone
from typing import Optional, Set
from zoneinfo import ZoneInfo

from kanbus.models import ProjectConfiguration, StandupConfiguration

MEETING_SCRIPT_PROFILE = "meeting-script"
DIRECTOR_BRIEF_PROFILE = "director-brief"
ROLLING_WINDOW = "rolling"
CALENDAR_WINDOW = "calendar"
DEFAULT_STANDUP_LOOKBACK = "24h"
DEFAULT_STANDUP_LOOKBACK_HOURS = 24
STANDUP_REPORT_TIME_ENV = "KANBUS_STANDUP_REPORT_TIME"
LOOKBACK_PATTERN = re.compile(r"^(\d+)([hd])$", re.IGNORECASE)


class StandupWindowError(ValueError):
    """Raised when standup window settings are invalid."""


def parse_rfc3339_timestamp(value: Optional[datetime | str]) -> Optional[datetime]:
    """Parse an RFC3339 timestamp into UTC.

    :param value: Timestamp value from issue or event data.
    :type value: Optional[datetime | str]
    :return: Parsed UTC datetime, or None when absent.
    :rtype: Optional[datetime]
    """
    if value is None:
        return None
    if isinstance(value, datetime):
        parsed = value
    else:
        parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


@dataclass(frozen=True)
class StandupWindowOverrides:
    """Optional CLI or API overrides for standup window settings.

    :param window: Optional window mode override.
    :type window: Optional[str]
    :param lookback: Optional lookback duration override.
    :type lookback: Optional[str]
    :param skip_weekends: Optional skip-weekends override.
    :type skip_weekends: Optional[bool]
    """

    window: Optional[str] = None
    lookback: Optional[str] = None
    skip_weekends: Optional[bool] = None


@dataclass(frozen=True)
class StandupWindowSettings:
    """Resolved standup window settings for report generation.

    :param window: Window mode (`rolling` or `calendar`).
    :type window: str
    :param lookback: Canonical lookback duration string.
    :type lookback: str
    :param lookback_hours: Lookback duration in hours.
    :type lookback_hours: int
    :param skip_weekends: Whether calendar mode bundles weekends on Monday.
    :type skip_weekends: bool
    :param timezone: Resolved standup timezone.
    :type timezone: ZoneInfo
    """

    window: str
    lookback: str
    lookback_hours: int
    skip_weekends: bool
    timezone: ZoneInfo


def parse_standup_lookback_hours(lookback: str) -> int:
    """Parse a standup lookback duration into hours.

    :param lookback: Duration string such as `24h` or `1d`.
    :type lookback: str
    :return: Lookback duration in hours.
    :rtype: int
    :raises StandupWindowError: When the duration format is invalid.
    """
    normalized = lookback.strip().lower()
    match = LOOKBACK_PATTERN.fullmatch(normalized)
    if match is None:
        raise StandupWindowError(f"invalid standup lookback: {lookback}")
    amount = int(match.group(1))
    unit = match.group(2)
    if amount <= 0:
        raise StandupWindowError(f"invalid standup lookback: {lookback}")
    if unit == "d":
        return amount * 24
    return amount


def format_standup_lookback(hours: int) -> str:
    """Format lookback hours as a canonical duration string.

    :param hours: Lookback duration in hours.
    :type hours: int
    :return: Canonical lookback string.
    :rtype: str
    """
    if hours % 24 == 0 and hours >= 24:
        days = hours // 24
        return f"{days}d"
    return f"{hours}h"


def resolve_standup_timezone(configuration: ProjectConfiguration) -> ZoneInfo:
    """Resolve the standup timezone from configuration.

    :param configuration: Project configuration.
    :type configuration: ProjectConfiguration
    :return: Resolved timezone.
    :rtype: ZoneInfo
    """
    timezone_name = configuration.standup.timezone
    if timezone_name:
        return ZoneInfo(timezone_name)
    local_timezone = datetime.now().astimezone().tzinfo
    if local_timezone is None:
        return ZoneInfo("UTC")
    if isinstance(local_timezone, ZoneInfo):
        return local_timezone
    return ZoneInfo(str(local_timezone))


def resolve_standup_window_settings(
    configuration: ProjectConfiguration,
    profile: Optional[str],
    overrides: Optional[StandupWindowOverrides] = None,
) -> StandupWindowSettings:
    """Resolve standup window settings from config, profile, and overrides.

    :param configuration: Project configuration.
    :type configuration: ProjectConfiguration
    :param profile: Optional standup profile identifier for profile defaults.
    :type profile: Optional[str]
    :param overrides: Optional CLI or API overrides.
    :type overrides: Optional[StandupWindowOverrides]
    :return: Resolved standup window settings.
    :rtype: StandupWindowSettings
    :raises StandupWindowError: When settings are invalid.
    """
    standup = configuration.standup
    window = standup.window
    lookback = standup.lookback
    skip_weekends = standup.skip_weekends

    if profile == MEETING_SCRIPT_PROFILE:
        window = CALENDAR_WINDOW
        skip_weekends = True
    elif profile == DIRECTOR_BRIEF_PROFILE:
        window = ROLLING_WINDOW
        lookback = DEFAULT_STANDUP_LOOKBACK
        skip_weekends = False

    if overrides is not None:
        if overrides.window is not None:
            window = overrides.window
        if overrides.lookback is not None:
            lookback = overrides.lookback
        if overrides.skip_weekends is not None:
            skip_weekends = overrides.skip_weekends

    if window not in {ROLLING_WINDOW, CALENDAR_WINDOW}:
        raise StandupWindowError(f"invalid standup window: {window}")

    lookback_hours = parse_standup_lookback_hours(lookback)
    return StandupWindowSettings(
        window=window,
        lookback=lookback,
        lookback_hours=lookback_hours,
        skip_weekends=skip_weekends,
        timezone=resolve_standup_timezone(configuration),
    )


def resolve_standup_report_time() -> datetime:
    """Resolve the standup report generation timestamp.

    :return: Report time in UTC.
    :rtype: datetime
    """
    override = os.environ.get(STANDUP_REPORT_TIME_ENV)
    if override:
        parsed = parse_rfc3339_timestamp(override)
        if parsed is None:
            raise StandupWindowError(
                f"invalid {STANDUP_REPORT_TIME_ENV}: {override}"
            )
        return parsed
    return datetime.now(timezone.utc)


def completed_calendar_dates(
    report_time: datetime,
    settings: StandupWindowSettings,
) -> Set[date]:
    """Return calendar dates that qualify for the completed bucket.

    :param report_time: Report generation time in UTC.
    :type report_time: datetime
    :param settings: Resolved standup window settings.
    :type settings: StandupWindowSettings
    :return: Qualifying local calendar dates.
    :rtype: Set[date]
    """
    local_report_time = report_time.astimezone(settings.timezone)
    report_day = local_report_time.date()
    if settings.skip_weekends and report_day.weekday() == 0:
        return {
            report_day - timedelta(days=3),
            report_day - timedelta(days=2),
            report_day - timedelta(days=1),
        }
    return {report_day - timedelta(days=1)}


def timestamp_calendar_date(
    timestamp: Optional[datetime | str],
    settings: StandupWindowSettings,
) -> Optional[date]:
    """Convert a timestamp to a local calendar date in standup timezone.

    :param timestamp: Timestamp to convert.
    :type timestamp: Optional[datetime | str]
    :param settings: Resolved standup window settings.
    :type settings: StandupWindowSettings
    :return: Local calendar date, or None when absent.
    :rtype: Optional[date]
    """
    parsed = parse_rfc3339_timestamp(timestamp)
    if parsed is None:
        return None
    return parsed.astimezone(settings.timezone).date()


def is_on_completed_calendar_day(
    timestamp: Optional[datetime | str],
    report_time: datetime,
    settings: StandupWindowSettings,
) -> bool:
    """Return whether a timestamp falls on a completed calendar day.

    :param timestamp: Timestamp to evaluate.
    :type timestamp: Optional[datetime | str]
    :param report_time: Report generation time in UTC.
    :type report_time: datetime
    :param settings: Resolved standup window settings.
    :type settings: StandupWindowSettings
    :return: True when the timestamp is on a qualifying calendar day.
    :rtype: bool
    """
    local_date = timestamp_calendar_date(timestamp, settings)
    if local_date is None:
        return False
    return local_date in completed_calendar_dates(report_time, settings)


def start_of_report_calendar_day(
    report_time: datetime,
    settings: StandupWindowSettings,
) -> datetime:
    """Return the UTC instant for the start of the report calendar day.

    :param report_time: Report generation time in UTC.
    :type report_time: datetime
    :param settings: Resolved standup window settings.
    :type settings: StandupWindowSettings
    :return: Start of report day in UTC.
    :rtype: datetime
    """
    local_report_time = report_time.astimezone(settings.timezone)
    start_local = datetime.combine(
        local_report_time.date(),
        time.min,
        tzinfo=settings.timezone,
    )
    return start_local.astimezone(timezone.utc)


def default_standup_configuration() -> StandupConfiguration:
    """Return default standup configuration values.

    :return: Default standup configuration.
    :rtype: StandupConfiguration
    """
    return StandupConfiguration()

File: rust/src/standup_window.rs
//! Standup time window resolution (rolling vs calendar).

use std::env;

use chrono::{DateTime, Datelike, Duration, TimeZone, Utc};
use chrono_tz::{Tz, UTC};
use regex::Regex;
use std::sync::OnceLock;

use crate::error::KanbusError;
use crate::models::{ProjectConfiguration, StandupConfiguration};
use crate::standup::{DIRECTOR_BRIEF_PROFILE, MEETING_SCRIPT_PROFILE};

pub const ROLLING_WINDOW: &str = "rolling";
pub const CALENDAR_WINDOW: &str = "calendar";
pub const DEFAULT_STANDUP_LOOKBACK: &str = "24h";
pub const STANDUP_REPORT_TIME_ENV: &str = "KANBUS_STANDUP_REPORT_TIME";

/// Optional CLI or API overrides for standup window settings.
#[derive(Debug, Clone, Default)]
pub struct StandupWindowOverrides {
    /// Optional window mode override.
    pub window: Option<String>,
    /// Optional lookback duration override.
    pub lookback: Option<String>,
    /// Optional skip-weekends override.
    pub skip_weekends: Option<bool>,
}

/// Resolved standup window settings for report generation.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StandupWindowSettings {
    /// Window mode (`rolling` or `calendar`).
    pub window: String,
    /// Canonical lookback duration string.
    pub lookback: String,
    /// Lookback duration in hours.
    pub lookback_hours: u32,
    /// Whether calendar mode bundles weekends on Monday.
    pub skip_weekends: bool,
    /// Resolved standup timezone.
    pub timezone: Tz,
}

fn lookback_pattern() -> &'static Regex {
    static PATTERN: OnceLock<Regex> = OnceLock::new();
    PATTERN.get_or_init(|| Regex::new(r"^(\d+)([hd])$").expect("lookback regex"))
}

/// Parse a standup lookback duration into hours.
///
/// # Errors
///
/// Returns `KanbusError::IssueOperation` when the duration format is invalid.
pub fn parse_standup_lookback_hours(lookback: &str) -> Result<u32, KanbusError> {
    let normalized = lookback.trim().to_ascii_lowercase();
    let captures = lookback_pattern().captures(&normalized).ok_or_else(|| {
        KanbusError::IssueOperation(format!("invalid standup lookback: {lookback}"))
    })?;
    let amount = captures
        .get(1)
        .and_then(|value| value.as_str().parse::<u32>().ok())
        .filter(|value| *value > 0)
        .ok_or_else(|| {
            KanbusError::IssueOperation(format!("invalid standup lookback: {lookback}"))
        })?;
    let unit = captures.get(2).map(|value| value.as_str()).unwrap_or("h");
    if unit == "d" {
        return Ok(amount * 24);
    }
    Ok(amount)
}

/// Resolve the standup timezone from configuration.
pub fn resolve_standup_timezone(configuration: &ProjectConfiguration) -> Tz {
    if let Some(timezone_name) = configuration.standup.timezone.as_deref() {
        if let Ok(timezone) = timezone_name.parse::<Tz>() {
            return timezone;
        }
    }
    let local_name = iana_time_zone::get_timezone().unwrap_or_else(|_| String::from("UTC"));
    local_name.parse::<Tz>().unwrap_or(UTC)
}

/// Resolve standup window settings from config, profile, and overrides.
///
/// # Errors
///
/// Returns `KanbusError::IssueOperation` when settings are invalid.
pub fn resolve_standup_window_settings(
    configuration: &ProjectConfiguration,
    profile: Option<&str>,
    overrides: &StandupWindowOverrides,
) -> Result<StandupWindowSettings, KanbusError> {
    let standup = &configuration.standup;
    let mut window = standup.window.clone();
    let mut lookback = standup.lookback.clone();
    let mut skip_weekends = standup.skip_weekends;

    if profile == Some(MEETING_SCRIPT_PROFILE) {
        window = CALENDAR_WINDOW.to_string();
        skip_weekends = true;
    } else if profile == Some(DIRECTOR_BRIEF_PROFILE) {
        window = ROLLING_WINDOW.to_string();
        lookback = DEFAULT_STANDUP_LOOKBACK.to_string();
        skip_weekends = false;
    }

    if let Some(override_window) = overrides.window.as_deref() {
        window = override_window.to_string();
    }
    if let Some(override_lookback) = overrides.lookback.as_deref() {
        lookback = override_lookback.to_string();
    }
    if let Some(override_skip_weekends) = overrides.skip_weekends {
        skip_weekends = override_skip_weekends;
    }

    if window != ROLLING_WINDOW && window != CALENDAR_WINDOW {
        return Err(KanbusError::IssueOperation(format!(
            "invalid standup window: {window}"
        )));
    }

    let lookback_hours = parse_standup_lookback_hours(&lookback)?;
    Ok(StandupWindowSettings {
        window,
        lookback,
        lookback_hours,
        skip_weekends,
        timezone: resolve_standup_timezone(configuration),
    })
}

/// Resolve the standup report generation timestamp.
///
/// # Errors
///
/// Returns `KanbusError::IssueOperation` when the override is invalid.
pub fn resolve_standup_report_time() -> Result<DateTime<Utc>, KanbusError> {
    if let Ok(override_value) = env::var(STANDUP_REPORT_TIME_ENV) {
        let normalized = override_value.replace('Z', "+00:00");
        let parsed = DateTime::parse_from_rfc3339(&normalized)
            .map_err(|_| {
                KanbusError::IssueOperation(format!(
                    "invalid {STANDUP_REPORT_TIME_ENV}: {override_value}"
                ))
            })?
            .with_timezone(&Utc);
        return Ok(parsed);
    }
    Ok(Utc::now())
}

/// Return calendar dates that qualify for the completed bucket.
pub fn completed_calendar_dates(
    report_time: DateTime<Utc>,
    settings: &StandupWindowSettings,
) -> Vec<chrono::NaiveDate> {
    let local_report_time = report_time.with_timezone(&settings.timezone);
    let report_day = local_report_time.date_naive();
    if settings.skip_weekends && report_day.weekday() == chrono::Weekday::Mon {
        return vec![
            report_day - Duration::days(3),
            report_day - Duration::days(2),
            report_day - Duration::days(1),
        ];
    }
    vec![report_day - Duration::days(1)]
}

fn timestamp_calendar_date(
    timestamp: Option<&DateTime<Utc>>,
    settings: &StandupWindowSettings,
) -> Option<chrono::NaiveDate> {
    timestamp.map(|value| value.with_timezone(&settings.timezone).date_naive())
}

/// Return whether a timestamp falls on a completed calendar day.
pub fn is_on_completed_calendar_day(
    timestamp: Option<&DateTime<Utc>>,
    report_time: DateTime<Utc>,
    settings: &StandupWindowSettings,
) -> bool {
    let Some(local_date) = timestamp_calendar_date(timestamp, settings) else {
        return false;
    };
    completed_calendar_dates(report_time, settings).contains(&local_date)
}

/// Return the UTC instant for the start of the report calendar day.
pub fn start_of_report_calendar_day(
    report_time: DateTime<Utc>,
    settings: &StandupWindowSettings,
) -> DateTime<Utc> {
    let local_report_time = report_time.with_timezone(&settings.timezone);
    let report_day = local_report_time.date_naive();
    settings
        .timezone
        .from_local_datetime(
            &report_day
                .and_hms_opt(0, 0, 0)
                .expect("midnight should be valid"),
        )
        .single()
        .map(|value| value.with_timezone(&Utc))
        .unwrap_or(report_time)
}

/// Return default standup configuration values.
pub fn default_standup_configuration() -> StandupConfiguration {
    StandupConfiguration::default()
}
